### ISSUE TITLE
[basic.compound] p4 What the definition of whether two objects are the same

### ISSUE THREAD

**Issue by xmh0511** [cplusplus/draft#5356]
[expr.static.cast] p13 states
> A prvalue of type “pointer to cv1 void” can be converted to a prvalue of type “pointer to cv2 T”, where T is an object type and cv2 is the same cv-qualification as, or greater cv-qualification than, cv1. ... Otherwise, if the original pointer value points to an object a, and there is an object b of type T (ignoring cv-qualification) that is **pointer-interconvertible with a**, the result is a pointer to b. Otherwise, the pointer value is unchanged by the conversion.

[basic.compound] p4.1 states
> Two objects a and b are pointer-interconvertible if:
>> - they are the same object, or
>> - [...]

Consider this example:
````cpp
int const v = 0;
void const* cvptr = &v;
void* vptr =  const_cast<void*>(cvptr);
int* iptr = static_cast<int*>(vptr );  // #3
````
The pointer value of `vptr` points to `v` as per [conv.ptr] p2, the result of  `const_cast<void*>(cvptr)` refers to the original entity(the same pointer value as vptr's) as per [expr.const.cast] p3. After figuring out what `vptr` is, we can pay more attention to `#3`, which is the case specified in [expr.static.cast] p13 where the conversion is that converts a prvalue of "pointer to void" to a prvalue of type "pointer to int". In this case, the original pointer value points to `v`(which is object `a`, whose type is `int const`) while the expected object `b` is of type `int`. They have different types. So, the issue is whether two objects that have **similar types but are not the same** can be the same object? From the perspective of English, when we say two objects are the same, they should be identical in any respect at least.

**Comment by jensmaurer** [comment]
`v` is a single object in this example. There are different ways to refer to that object, e.g. by saying `v` or by saying `*iptr`. Nothing to see here. (Yes, modifying `*iptr` is undefined behavior because the object designated by `v` is a const object, but that's not directly relevant here.)

**Comment by jensmaurer** [comment]
[expr.static.cast] p13 says:

> Otherwise, if the original pointer value points to an object a, and there is an object b of type T (ignoring cv-qualification) that is pointer-interconvertible (6.8.3) with a, the result is a pointer to b. Otherwise, the pointer value is unchanged by the conversion.

We agree the original pointer points to the v object, of type "const int". And there is an object b of type "int" (ignoring cv-qualification), which is the same object as the v object, so the result is a pointer to that v object. In this example, there is only ever one object we're talking about, and I think it goes without definition that x is the same as x (for any x).

**Comment by xmh0511** [comment]
> And there is an object b of type "int" (ignoring cv-qualification), which is the same object as the v object

How could be an object of type `int` the same as an object of type `const int`? If it were, then `iptr` would point to an object of type int
> the result is a pointer to **b**. 

 which means, any modification to that object is well-defined, because `b` is not a const object, as per [dcl.type.cv] p4

**Comment by jensmaurer** [comment]
"is exactly that b and a are not the same object"

That can't be, because there is only a single object. Either "b" doesn't exist at all ("there is an object b of type T"), or it exists and then it's the same as a.

**Comment by frederick-vs-ja** [comment]
I don't see anything here creating a distinct `int` object. It is **two different lvalue expressions** `v` and `*iptr` that denote **the same object** of type `const int`, although `*iptr` is of type `int` (`decltype(*iptr)` is `int&`).

**Comment by xmh0511** [comment]
@frederick-vs-ja  The key point is whether the conversion falls into 
> Otherwise, if the original pointer value points to an object a, and there is an object b of type T (ignoring cv-qualification) that is [pointer-interconvertible](https://eel.is/c++draft/basic.compound#def:pointer-interconvertible) with a, the result is a pointer to b[.](https://eel.is/c++draft/expr.static.cast#13.sentence-3)

or
> Otherwise, the pointer value is unchanged by the conversion[.](https://eel.is/c++draft/expr.static.cast#13.sentence-4)

**Comment by frederick-vs-ja** [comment]
> you seem to admit that the conversion should fall into the second branch.

When there is not a different _b_ object, IIUC two branches are not distinguishable. Note the "ignoring cv-qualification".

> Take more attention to [expr.unary.op] p1
> 
> > The unary * operator performs [indirection](https://eel.is/c++draft/expr.unary#def:indirection): the expression to which it is applied shall be a pointer to an object type, or a pointer to a function type and the result is an lvalue referring to **the object** or function to which the **expression points**[.](https://eel.is/c++draft/expr.unary#op-1.sentence-1)

"An lvalue referring to the object" is not needed to have the type of that object.

**Comment by xmh0511** [comment]
> Note the "ignoring cv-qualification".

The destination of this conversion is "prvalue of type pointer to int(T)" where `T` is just the unqualified-version type, there is no top-level cv-qualification. And thus, `b` is said to be an object of type `int`(again, there is no cv-qualification to be ignored). 

> "An lvalue referring to the object" is not needed to have the type of that object.

An lvalue certainly can have a different type as the type of the actual object to which the expression points. The correct interpretation to [expr.unary.op] p1 is specified in [P1839](https://wg21.link/p1839). As I have said in the above comments, the original object `a` is the object denoted by `v`, which is an object of type `const int`, which is created by the definition of `v`. The `b` here is specified to an object of type `int`. Any same object cannot both have different types. 

> two branches are not distinguishable

No, since `b` is an object of type `int`, if the first branch works, the result is a pointer to **b**, which is an object of type **int**.

**Comment by frederick-vs-ja** [comment]
> again, there is no cv-qualification to be ignored

IIUC there is, as "ignoring cv-qualification" is applied to the phrase "there is an object _b_ of type `T`", which means that both the cv-qualification of the type of _b_ and that of `T` are ignored for such purpose.

**Comment by xmh0511** [comment]
> which means that both the cv-qualification of the type of b and that of T are ignored for such purpose.

I don't understand what it means. Do you select to directly ignore the introduction of `b`? 

> there is **an object b of type T**

Could you point out what `T` is in the complete rule? `T` is introduced in this sentence

> A prvalue of type “pointer to cv1 void” can be converted to a prvalue of type “pointer to cv2 T”, where T is an object type

What is the destination of the conversion? In this case, it is the prvalue of type pointer to cv2 int, Isn't it? So, `T` is obviously **int**. Do you agree with that?

**Comment by frederick-vs-ja** [comment]
> Do you select to directly ignore the introduction of `b`?

No. How can I select to directly ignore the introduction of `b` while I paste it?

> Isn't `T` is exactly the specified type of `b`?

This is what I disagree. Given `T` being `int` in the example, my comprehension is that the condition "there is an object _b_ of type `T` (ignoring cv-qualification)" is transformed into "there is an object _b_ whose type is `int`, `const int`, `volatile int`, or `const volatile int`".

> So, `T` is obviously `int`. Do you agree with that?

Totally agree for this.

**Comment by xmh0511** [comment]
>  my comprehension is that the condition "there is an object b of type T (ignoring cv-qualification)" is transformed into "there is an object b whose type is int, const int, volatile int, or const volatile int".

That is our divergence here. In other words, the ambiguity of this portion. I construe "ignore cv-qualification" for any type `T` to ignore the top-level cv-qualification of `T`. Specifically, for any type `T`, its qualification-decomposition is `cv U`(where n = 0), we only take `U`, this is "ignoring cv-qualification". 

If it were the meaning of what you said, we should take the same manner as [over.match.funcs.general] p7 did
> for any permissible type cv U, any cv2 U, cv2 U&, or cv2 U&& is also a permissible type[.](https://eel.is/c++draft/over.match.funcs#general-7.sentence-4)

Instead of saying ignore the cv-qualification.

**Comment by languagelawyer** [comment]
> I construe "ignore cv-qualification" for any type `T` to ignore the top-level cv-qualification of `T`

When the standard compares 2 types and says «(ignoring cv-qualification)», in all places it means to ignore cv-qualification on both types, not only on that which is close to «(ignoring cv-qualification)».

**Comment by languagelawyer** [comment]
> That means the result would point to object `b`, which is of type `int`(the rule specifies the type of that object)

Where?

**Comment by xmh0511** [comment]
> there is an object b of type T

>  there is an object b of type T

T is the type specified in  pointer to cv2 T, where the top-level cv-qualification of the pointed-to type is denoted by `cv2`, thus `T` is a cv-unqualified type.
